tempesta_fw/http_parser.c

```diff
@@ -2428,274 +2428,237 @@ __req_parse_referer(TfwHttpMsg *hm, unsigned char *data, size_t len)
-	__FSM_STATE(I_DateYear) {
-		__fsm_sz = __data_remain(p);
-		__fsm_n = parse_int_ws(p, __fsm_sz, &parser->_acc);
-		if (__fsm_n == CSTR_POSTPONE) {
-			parser->_i_st = &&I_DateYear;
-			__msg_hdr_chunk_fixup(data, len);
+	__FSM_STATE(I_WDate4) {
+		if (likely(c == ',')) {
+			parser->date_fmt = &rfc822_fmt;
+			__FSM_I_MOVE(I_WDateSP);
 		}
-		if (__fsm_n < 0)
-			return __fsm_n;
-		parser->_date = __year_day_secs(parser->_acc, parser->_date);
-		if (parser->_date < 0)
-			return CSTR_NEQ;
-		parser->_acc = 0;
-		__FSM_I_MOVE_n(I_DateHourSP, __fsm_n);
-	}
-
-	__FSM_STATE(I_DateHourSP) {
-		if (likely(c == ' '))
-			__FSM_I_MOVE(I_DateHour);
-		return CSTR_NEQ;
-	}
-
-	__FSM_STATE(I_DateHour) {
-		__fsm_sz = __data_remain(p);
-		__fsm_n = parse_int_a(p, __fsm_sz, colon_a, &parser->_acc);
-		if (__fsm_n == CSTR_POSTPONE) {
-			parser->_i_st = &&I_DateHour;
-			__msg_hdr_chunk_fixup(data, len);
+		if ('a' <= c && c <= 'z') {
+			parser->date_fmt = &rfc850_fmt;
+			__FSM_I_MOVE(I_WDate5);
+		}
+		if (c == ' ') {
+			parser->date_fmt = &isoc_fmt;
+			__FSM_I_MOVE(I_Loop);
 		}
-		if (__fsm_n < 0)
-			return __fsm_n;
-		parser->_date += parser->_acc * 3600;
-		parser->_acc = 0;
-		__FSM_I_MOVE_n(I_DateMinCln, __fsm_n);
-	}
-
-	__FSM_STATE(I_DateMinCln) {
-		if (likely(c == ':'))
-			__FSM_I_MOVE(I_DateMin);
 		return CSTR_NEQ;
 	}
 
-	__FSM_STATE(I_DateMin) {
-		__fsm_sz = __data_remain(p);
-		__fsm_n = parse_int_a(p, __fsm_sz, colon_a, &parser->_acc);
-		if (__fsm_n == CSTR_POSTPONE) {
-			parser->_i_st = &&I_DateMin;
-			__msg_hdr_chunk_fixup(data, len);
+	__FSM_STATE(I_WDate5) {
+		if ('a' <= c && c <= 'z') {
+			__FSM_I_MOVE(I_WDate5);
+		}
+		if (c == ',') {
+			__FSM_I_MOVE(I_WDateSP);
 		}
-		if (__fsm_n < 0)
-			return __fsm_n;
-		parser->_date += parser->_acc * 60;
-		parser->_acc = 0;
-		__FSM_I_MOVE_n(I_DateSecCln, __fsm_n);
-	}
-
-	__FSM_STATE(I_DateSecCln) {
-		if (likely(c == ':'))
-			__FSM_I_MOVE(I_DateSec);
 		return CSTR_NEQ;
 	}
 
-	__FSM_STATE(I_DateSec) {
-		__fsm_sz = __data_remain(p);
-		__fsm_n = parse_int_ws(p, __fsm_sz, &parser->_acc);
-		if (__fsm_n == CSTR_POSTPONE) {
-			parser->_i_st = &&I_DateSec;
-			__msg_hdr_chunk_fixup(data, len);
+	__FSM_STATE(I_WDateSP) {
+		if (likely(c == ' ')) {
+			parser->_acc = 0;
+			__FSM_I_MOVE(I_Loop);
 		}
-		if (__fsm_n < 0)
-			return __fsm_n;
-		parser->_date += parser->_acc;
-		parser->_acc = 0;
-		__FSM_I_MOVE_n(I_DateSecSP, __fsm_n);
-	}
-
-	__FSM_STATE(I_DateSecSP) {
-		if (likely(c == ' '))
-			__FSM_I_MOVE(I_DateZone);
 		return CSTR_NEQ;
 	}
 
-	__FSM_STATE(I_DateZone) {
-		TRY_STR("gmt", I_DateZone, I_EoL);
-		TRY_STR_INIT();
-		return CSTR_NEQ;
+	__FSM_STATE(I_Loop) {
+		TfwHttpDateFmt *fmt = parser->date_fmt;
+
+		if (parser->_acc >= fmt->len) {
+			time_t date = __parse_date(parser->date_str, fmt);
+			if (date < 0)
+				return CSTR_NEQ;
+			parser->_date = date;
+			__FSM_I_JMP(I_EoL);
+		}
+		if (c < fmt->min[parser->_acc] || c > fmt->max[parser->_acc]) {
```

Comment:
This is a possible point for further optimizations. However, it's not immediately clear if vectorization will gain anything here.

Reply:
Yeah, I was talking earlier about compilers already doing the optimization, and about why one shouldn't do that by hand, since it makes code harder to understand at a first glance. But in this particular case compiler skips the optimize. (We use GCC 6.3, but it's the same for more recent versions too). Even simple
```c
void func_simple2(char c, char min, char max) {
    if (min < c && c < max)
        puts("check");
}
```
are not optimized.
Turned out, the reason is in unpredictable `min` and `max` values. I assumed that `min <= max` just from the names. It's a reasonable assumption for a human, but a compiler cannot afford that. So since in the case of `min > max` optimization breaks the control flow, it's not used.

So, rule of thumb: if borders are immediate constants, two comparisons will likely be optimized to one. Otherwise, a manual optimization is required.

Reply:
Thanks @i-rinat . This means that we do need the hacker's delight trick here, but don't need it in comparisons like `'a' <= c && c <= 'z'`. @avbelov23 please make the optimization and leave a comment about compiler optimizations and why we do need manual optimization here.